third_party/nvidia/test/NVWS/ops.mlir

```diff
@@ -0,0 +1,94 @@
+    %0 = nvws.aref.create %d, %e : !nvws.aref<[!ttg.memdesc<1x64x16xf16, #shared0, #smem>, !ttg.memdesc<1x16x32xf16, #shared0, #smem>]>
+    nvws.aref.put %0 as (%b0 : !ttg.memdesc<1x64x16xf16, #shared0, #smem>, %b1 : !ttg.memdesc<1x16x32xf16, #shared0, #smem>) {
+      nvws.aref.return
+    } : (!nvws.aref<[!ttg.memdesc<1x64x16xf16, #shared0, #smem>, !ttg.memdesc<1x16x32xf16, #shared0, #smem>]>) -> ()
+    tt.return
+  }
+}
+
+// -----
+
+#shared0 = #ttg.nvmma_shared<{swizzlingByteWidth = 32, transposed = false, elementBitWidth = 16}>
+#smem = #ttg.shared_memory
+module attributes {"ttg.target" = "cuda:0", "ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 1 : i32, "ttg.threads-per-warp" = 32 : i32} {
+  // CHECK-LABEL: aref_create_batch
+  // CHECK: nvws.aref.create
+  tt.func @aref_create_batch(%d : !ttg.memdesc<1x64x16xf16, #shared0, #smem>, %e : !ttg.memdesc<1x16x32xf16, #shared0, #smem>) {
+    %0 = nvws.aref.create %d, %e : !nvws.aref<[!ttg.memdesc<1x64x16xf16, #shared0, #smem>, !ttg.memdesc<1x16x32xf16, #shared0, #smem>], 1>
+    tt.return
+  }
+
+}
+
+// -----
+
+#shared0 = #ttg.nvmma_shared<{swizzlingByteWidth = 32, transposed = false, elementBitWidth = 16}>
+#smem = #ttg.shared_memory
+module attributes {"ttg.target" = "cuda:0", "ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 1 : i32, "ttg.threads-per-warp" = 32 : i32} {
+  // CHECK-LABEL: aref_get_batch
+  // CHECK: nvws.aref.get
+  tt.func @aref_get_batch(%d : !ttg.memdesc<1x64x16xf16, #shared0, #smem>, %e : !ttg.memdesc<1x16x32xf16, #shared0, #smem>) {
+    %c0_i32 = arith.constant {ttg.partition = [0, 1]} 0 : i32
+    %0 = nvws.aref.create %d, %e : !nvws.aref<[!ttg.memdesc<1x64x16xf16, #shared0, #smem>, !ttg.memdesc<1x16x32xf16, #shared0, #smem>], 1>
+    %1 = nvws.aref.get %0[%c0_i32] as (%b0 : !ttg.memdesc<64x16xf16, #shared0, #smem>, %b1 : !ttg.memdesc<16x32xf16, #shared0, #smem>) {
+      nvws.aref.return %b0 : !ttg.memdesc<64x16xf16, #shared0, #smem>
+    } : (!nvws.aref<[!ttg.memdesc<1x64x16xf16, #shared0, #smem>, !ttg.memdesc<1x16x32xf16, #shared0, #smem>], 1>, i32) -> !ttg.memdesc<64x16xf16, #shared0, #smem>
+    tt.return
+  }
+}
+
+// -----
+
+#shared0 = #ttg.nvmma_shared<{swizzlingByteWidth = 32, transposed = false, elementBitWidth = 16}>
+#smem = #ttg.shared_memory
+module attributes {"ttg.target" = "cuda:0", "ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 1 : i32, "ttg.threads-per-warp" = 32 : i32} {
+  // CHECK-LABEL: aref_put_batch
+  // CHECK: nvws.aref.put
+  tt.func @aref_put_batch(%d : !ttg.memdesc<1x64x16xf16, #shared0, #smem>, %e : !ttg.memdesc<1x16x32xf16, #shared0, #smem>) {
+    %c0_i32 = arith.constant {ttg.partition = [0, 1]} 0 : i32
+    %0 = nvws.aref.create %d, %e : !nvws.aref<[!ttg.memdesc<1x64x16xf16, #shared0, #smem>, !ttg.memdesc<1x16x32xf16, #shared0, #smem>], 1>
+    nvws.aref.put %0[%c0_i32] as (%b0 : !ttg.memdesc<64x16xf16, #shared0, #smem>, %b1 : !ttg.memdesc<16x32xf16, #shared0, #smem>) {
```

Comment:
I remember we discussed this, but what are the semantics for `nvws.aref.put` with an in-register type like a tensor or scalar? With memrefs, you write to the memdesc, but it's not clear what the semantics are for a tensor.

Reply:
To return it from the put, I'll add it to the doc string.

Reply:
And add a test, I just realized these are all memdesc

Reply:
Gotcha, but is the block argument still there for these types? What happens if you try to return a memdesc?

Reply:
for put, we can have the following 
```
%val = ..
aref.put %aref[%idx] as %buf {
   local_store %val, %buf
      or
   tmem_store %val, %buf
}
```

Reply:
That makes sense for memdesc values, but what about in-register tensors?

Reply:
not sure I follow, can you give an example?

Reply:
For example, what is the difference between how `aref<memdesc<MxNxdtype>>` are `aref<tensor<MxNxdtype>>` are used with these two ops? I understand that the memdesc is used with explicit loads and stores, but what about the latter?

Reply:
I just pushed better verification for nvws.aref.return + tests and improved documentation, please let me know if that clears it up.

Reply:
we don't have `aref<tensor<MxNxdtype>>`, we only have `aref<memdesc<MxNxdtype>>`. To move tensor to another warp group, the region inside `aref_put` will need to copy tensor to buffer, and `aref_get` need to copy it back, copies are explicit, and user decide how to move tensors to other.

Reply:
The testcase below has

```mlir
    %0 = nvws.aref.create %d, %e : !nvws.aref<[tensor<1x64x16xf16>, tensor<1x16x32xf16>], 1>
    nvws.aref.put %0[%c0_i32] as (%b0 : tensor<64x16xf16>, %b1 : tensor<16x32xf16>) {
      %1 = math.exp %b0 : tensor<64x16xf16>
      %2 = math.cos %b1 : tensor<16x32xf16>
      nvws.aref.return %1, %2 : tensor<64x16xf16>, tensor<16x32xf16>
    } : (!nvws.aref<[tensor<1x64x16xf16>, tensor<1x16x32xf16>], 1>, i32) -> ()
```

It appears that the tensor expects to be returned explicitly, but what are the semantics of accessing the block argument? cc @mbrookhart

Reply:
:/ I guess this isn't quite clear, you're right. I should probably put the as (slices) bit at the end, it probably doesn't need to be block args for put.